[Keith Png] iP

KeithPJX opened the pull request:
### Gibson

> "When life gives you lemons? Don't make lemonade. Make life take the lemons back! Get mad! 'I don't want your damn lemons! What am I supposed to do with these?" - Cave Johnson ([source](https://theportalwiki.com/wiki/Cave_Johnson_voice_lines))

Gibson frees your mind of having to remember things you need to do. It's,

- text-based (for now)
- easy to learn
- ~~FAST~~ _REALLY FAST_ to use

All you need to do is,

1. download it from [here](https://github.com/KeithPJX/ip/releases/tag/A-Jar)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :grin:

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon
- [ ] Reminders (coming soon)
-------------------------
If you're a Java programmer, you can use it to practice Java too. Here's the `main` method:
``` Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

KeithPJX committed:
Change Duke to Gibson

KeithPJX committed:
Add echo and exit functions

KeithPJX committed:
Add add and list functions

KeithPJX committed:
Add Task class

KeithPJX committed:
Fix tasks not allowing spaces

KeithPJX committed:
Add mark and unmark function

KeithPJX committed:
Task.java: Add "[T]" in toString() method

KeithPJX committed:
Add Deadline and Event classes

KeithPJX committed:
Gibson.java: Add todo command

KeithPJX committed:
Gibson.java: Add deadline function

KeithPJX committed:
Gibson.java: Add event function

KeithPJX committed:
Gibson.java: Change command not recognized message

KeithPJX committed:
Add test in /text-ui-test/

KeithPJX committed:
Add numerous exception handling throughout the program

KeithPJX committed:
Gibson.java: Fix bug with subStringBeforeAfterToken() introduced in previous commit

KeithPJX committed:
Add delete function

KeithPJX committed:
Fix IndexOutOfBounds exceptions with mark, unmark, and delete functions

KeithPJX committed:
Add /data/ to .gitignore

KeithPJX committed:
Add save function

KeithPJX committed:
Add load function and fix minor bugs

KeithPJX committed:
Add date and time recognition in deadline command

KeithPJX committed:
Merge branch 'branch-Level-7'

KeithPJX committed:
Merge branch 'branch-Level-8'

KeithPJX committed:
Add storage class

KeithPJX committed:
Rename RegexUtility to Parser

KeithPJX committed:
Fix bug in Deadline

KeithPJX committed:
Add Ui class

KeithPJX committed:
Fix format error

KeithPJX committed:
Package .java files

KeithPJX committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

KeithPJX committed:
Add JUnit tests for Parser.java and TaskList.java

KeithPJX committed:
Add JavaDoc

KeithPJX committed:
Change code format

KeithPJX committed:
Add find function

KeithPJX committed:
Merge branch 'branch-A-CodingStandard'

KeithPJX committed:
Merge branch 'branch-A-JavaDoc'

KeithPJX committed:
Merge branch 'branch-Level-9'

KeithPJX committed:
Fix error message in find function

KeithPJX committed:
Add checkstyle

KeithPJX committed:
Change code to meet given coding standard

KeithPJX committed:
Add basic GUI

KeithPJX committed:
Restructure files

KeithPJX committed:
Edit JavaDocs and fix ordering of imports

KeithPJX committed:
Add GUI using FXML

cliftonfelix commented on src/main/java/gibson/Ui.java:
```diff
@@ -0,0 +1,43 @@
+import java.io.IOException;
+
+import gibson.task.Parser;
+import gibson.view.MainWindow;
```
Isn't it better to put a new line in between to separate the 2 chunks of import statements?

cliftonfelix commented on src/main/java/gibson/Storage.java:
```diff
@@ -0,0 +1,76 @@
+     */
+    public Storage(String directoryString, String fileName) {
+        this.directoryPath = Paths.get(directoryString);
+        this.filePath = Paths.get(directoryString, fileName);
```
For line 27 and 28, isn't "this." unnecessary?
IIRC in 2103, it's under an advanced rule so it's optional to follow

cliftonfelix commented on src/main/java/gibson/task/Parser.java:
```diff
@@ -0,0 +1,191 @@
+        } else if (input.equals("list")) {
+            return "Here are the task(s) in your list:\n" + taskList.toString();
+            // MARK
+        } else if (Pattern.matches("mark [0-9]+", input)) {
```
I like the idea of using Regex here

cliftonfelix commented on src/main/java/gibson/task/Parser.java:
```diff
@@ -0,0 +1,191 @@
+        } else {
+            return "I'm sorry. I do not know what that means.";
+        }
+    }
```
processInput method seems too long for me. 

As per 2103 Code Quality guide:

> Be wary when a method is longer than the computer screen, and take corrective action when it goes beyond 30 LOC (lines of code). The bigger the haystack, the harder it is to find a needle. ([source](https://nus-cs2103-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html#guideline-maximise-readability:~:text=Be%20wary%20when%20a%20method%20is%20longer%20than%20the%20computer%20screen%2C%20and%20take%20corrective%20action%20when%20it%20goes%20beyond%2030%20LOC%20(lines%20of%20code).%20The%20bigger%20the%20haystack%2C%20the%20harder%20it%20is%20to%20find%20a%20needle.))

Perhaps 1 thing that you can do is to put try catch inside each case clause to a separate method or class.

cliftonfelix commented on src/main/java/gibson/task/Parser.java:
```diff
@@ -0,0 +1,191 @@
+            throw new IllegalArgumentException("Unable to find token in string");
+        }
+    }
+}
```
Shouldn't `getTrailingInt`, `substringAfterToken`, `substringBeforeAfterToken` methods have `throws IllegalArgumentException`?

cliftonfelix reviewed (commented):
Overall, code structure and quality LGTM, just few nits.

Aishwarya-Hariharan-Iyer commented on src/main/java/gibson/task/Deadline.java:
```diff
@@ -0,0 +1,52 @@
+ */
+public class Deadline extends Task {
+
+    private static final String[] FORMATS = new String[] {
```
I really like how you have all the accepted input formats in one place/array! Perhaps you could also consider using or maybe adding predefined [DateTimeFormatter constants](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#ISO_LOCAL_DATE) from the Java API?

Aishwarya-Hariharan-Iyer commented on src/main/java/gibson/task/Task.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Override
+    public String toString() {
+        if (done) {
```
I find this piece of code very neat! May I ask why you chose to `getChar()` at the level of Tasks, instead of having the subclasses print out their respective letters in their `toString()` methods?

Aishwarya-Hariharan-Iyer commented on src/main/java/gibson/task/Parser.java:
```diff
@@ -0,0 +1,191 @@
+                return "There is not task numbered as " + number + ".";
+            }
+            // UNMARK
+        } else if (Pattern.matches("unmark [0-9]+", input)) {
```
I find it to be a really well-thought-out piece of code! Perhaps, you could consider storing the commands as constants or as an array?

Aishwarya-Hariharan-Iyer commented on src/main/java/gibson/task/Parser.java:
```diff
@@ -0,0 +1,191 @@
+     * @throws IllegalArgumentException if token is not found in the string
+     */
+    public static String substringAfterToken(String string, String token) {
+        Pattern pattern = Pattern.compile(token);
```
Considering it is a method, do you think it would be better to perhaps name it using a verb as per the given [coding standards](https://se-education.org/guides/conventions/java/intermediate.html)? The same goes for other such methods as well.

Aishwarya-Hariharan-Iyer commented on src/main/java/gibson/task/Task.java:
```diff
@@ -0,0 +1,46 @@
+public class Task {
+    private String taskString;
+    private boolean done;
+
```
Considering it is a boolean, do you think naming it as `isDone` (or something similar) would be more in accordance with the [coding standards](https://se-education.org/guides/conventions/java/intermediate.html)? The same goes for other such boolean names as well.

Aishwarya-Hariharan-Iyer commented on src/main/java/gibson/task/Task.java:
```diff
@@ -0,0 +1,46 @@
+ */
+public class Task {
+    private String taskString;
+    private boolean done;
```
Additionally, though I find the use of booleans more elegant, do you think that perhaps using integer representations (0=not done, 1 = done, 2=in progress, etc.) makes the code more extendable, given we can add more 'status of completion' values?

Aishwarya-Hariharan-Iyer reviewed (commented):
LGTM. Just a few changes related to naming conventions and a few suggestions/questions.

KeithPJX committed:
Update build.gradle

KeithPJX committed:
Add assertion in Deadline.java

Construction of Deadline object will not work if FORMATS array is
empty.

Adding an assertion ensures developers have at least 1 element in
the FORMATS array if they were to change it.

KeithPJX committed:
Add check to see if there is data to load

This was in the program previously but was removed accidentally
when moving the code around.

Adding it back will stop the calling of the load() method in
Storage.java when there is nothing to load.

KeithPJX committed:
Refactor processInput method in Parser.java

The method processInput was really long and hard to read.

Applying SLAP on the method makes it more readable.

KeithPJX committed:
Merge pull request #1 from KeithPJX/branch-A-Assertions

Add assertion in Deadline.java

KeithPJX committed:
Merge pull request #2 from KeithPJX/branch-A-CodeQuality

Refactor Parser.java

KeithPJX committed:
Refactor all three add task commands

A few lines of code was repeated across three methods.

Removing the duplicated code and putting it in its own method will
fix this violation of DRY.

KeithPJX committed:
Add duplicate detection function

Duplicate tasks can be added.

Adding a check for duplicates before adding a task to the system
will solve the problem.

KeithPJX committed:
Merge branch 'branch-C-DetectDuplicates'
